File: examples/erlang/04-control-flow/main.erl
%%% Control Flow in Erlang
%%% ======================
%%% Erlang has no loops. Control flow via pattern matching,
%%% case/if expressions, guards, and recursion.

-module(main).
-export([start/0]).

start() ->
    %% ── case expression ──────────────────────────

    io:format("--- case expression ---~n"),
    io:format("describe(42) = ~s~n", [describe(42)]),
    io:format("describe(ok) = ~s~n", [describe(ok)]),
    io:format("describe(\"hello\") = ~s~n", [describe("hello")]),

    %% ── if expression (guards only) ──────────────

    io:format("~n--- if expression ---~n"),
    io:format("grade(95) = ~s~n", [grade(95)]),
    io:format("grade(82) = ~s~n", [grade(82)]),
    io:format("grade(45) = ~s~n", [grade(45)]),

    %% ── Recursion (looping via recursion) ────────

    io:format("~n--- recursion (countdown) ---~n"),
    countdown(5),

    io:format("~n--- recursion (sum 1..10) ---~n"),
    io:format("sum(10) = ~p~n", [sum(10)]),

    io:format("~n--- recursion (factorial) ---~n"),
    io:format("factorial(5) = ~p~n", [factorial(5)]),

    %% ── Tail-recursive loop ──────────────────────

    io:format("~n--- tail recursion ---~n"),
    io:format("powers_of_two(8) = ~p~n", [powers_of_two(8)]),

    %% ── Higher-order iteration ───────────────────

    io:format("~n--- lists:foreach ---~n"),
    lists:foreach(fun(X) -> io:format("item: ~p~n", [X]) end, [1, 2, 3]),

    io:format("~n--- lists:map ---~n"),
    Squares = lists:map(fun(X) -> X * X end, [1, 2, 3, 4, 5]),
    io:format("squares = ~p~n", [Squares]),

    io:format("~n--- lists:filter ---~n"),
    Evens = lists:filter(fun(X) -> X rem 2 == 0 end, [1, 2, 3, 4, 5, 6]),
    io:format("evens = ~p~n", [Evens]),

    %% ── List comprehensions ──────────────────────

    io:format("~n--- list comprehensions ---~n"),
    Doubles = [X * 2 || X <- [1, 2, 3, 4, 5]],
    io:format("doubles = ~p~n", [Doubles]),

    Odds = [X || X <- [1, 2, 3, 4, 5, 6, 7, 8, 9, 10], X rem 2 == 1],
    io:format("odds = ~p~n", [Odds]),

    %% ── Early exit via pattern matching ──────────

    io:format("~n--- early return (find first) ---~n"),
    io:format("find_even([1,3,5,8,9]) = ~p~n", [find_even([1, 3, 5, 8, 9])]),

    %% ── Summary ──────────────────────────────────

    io:format("~n--- Summary ---~n"),
    io:format("Erlang control flow:~n"),
    io:format("  - case: pattern matching on values~n"),
    io:format("  - if: guard-based branching~n"),
    io:format("  - Recursion: primary looping mechanism~n"),
    io:format("  - Tail recursion: constant stack space~n"),
    io:format("  - HOFs: lists:map, filter, foreach~n"),
    io:format("  - List comprehensions: [Expr || X <- List, Guard]~n"),
    ok.

%% ── case expression ────────────────────────────────

describe(X) when is_integer(X) -> "integer";
describe(X) when is_atom(X)    -> "atom";
describe(X) when is_list(X)    -> "string/list";
describe(_)                    -> "other".

%% ── if expression (guards only) ────────────────────

grade(Score) when Score >= 90 -> "A";
grade(Score) when Score >= 80 -> "B";
grade(Score) when Score >= 70 -> "C";
grade(Score) when Score >= 60 -> "D";
grade(_)                      -> "F".

%% ── Simple recursion ───────────────────────────────

countdown(0) ->
    io:format("blastoff!~n");
countdown(N) when N > 0 ->
    io:format("~p...~n", [N]),
    countdown(N - 1).

%% ── Tail-recursive sum ─────────────────────────────

sum(N) -> sum(N, 0).
sum(0, Acc) -> Acc;
sum(N, Acc) when N > 0 -> sum(N - 1, Acc + N).

factorial(0) -> 1;
factorial(N) when N > 0 -> N * factorial(N - 1).

%% ── Tail-recursive powers of two ───────────────────

powers_of_two(N) -> powers_of_two(N, 1, []).
powers_of_two(0, _, Acc) -> lists:reverse(Acc);
powers_of_two(N, Cur, Acc) when N > 0 ->
    powers_of_two(N - 1, Cur * 2, [Cur | Acc]).

%% ── Find first even (early exit via matching) ──────

find_even([]) -> none;
find_even([H | _]) when H rem 2 == 0 -> H;
find_even([_ | T]) -> find_even(T).

File: examples/haskell/04-control-flow/Main.hs
{-
  Control Flow in Haskell
  =======================
  Haskell is expression-oriented: no statements, only expressions.
  Branching via if/then/else, guards, and case. No loops — use
  recursion or higher-order functions instead.
-}

module Main where

-- ── Helper ──────────────────────────────────────────

section :: String -> IO ()
section title = putStrLn ("\n--- " ++ title ++ " ---")

-- ── if/then/else ────────────────────────────────────

classify :: Int -> String
classify n =
    if n < 0 then "negative"
    else if n == 0 then "zero"
    else "positive"

-- ── Guards ──────────────────────────────────────────

grade :: Int -> Char
grade score
    | score >= 90 = 'A'
    | score >= 80 = 'B'
    | score >= 70 = 'C'
    | score >= 60 = 'D'
    | otherwise   = 'F'

-- ── case expression ─────────────────────────────────

describe :: Either Int String -> String
describe value = case value of
    Left n  -> "integer: " ++ show n
    Right s -> "string: " ++ s

-- ── Recursion (countdown) ───────────────────────────

countdown :: Int -> IO ()
countdown 0 = putStrLn "blastoff!"
countdown n = do
    putStrLn (show n ++ "...")
    countdown (n - 1)

-- ── Tail recursion (sum) ────────────────────────────

sumUpTo :: Int -> Int
sumUpTo n = go n 0
  where
    go 0 acc = acc
    go k acc = go (k - 1) (acc + k)

-- ── Higher-order iteration ──────────────────────────

printSquares :: [Int] -> IO ()
printSquares nums = mapM_ (\x -> putStrLn ("square of " ++ show x ++ " = " ++ show (x * x))) nums

-- ── List comprehensions ─────────────────────────────

-- Generate even numbers from 1 to 20
evens :: [Int]
evens = [x | x <- [1..20], x `mod` 2 == 0]

-- Pythagorean triples
pythag :: [(Int, Int, Int)]
pythag = [(a, b, c) | c <- [1..20], b <- [1..c], a <- [1..b], a*a + b*b == c*c]

-- ── Main ────────────────────────────────────────────

main :: IO ()
main = do
    putStrLn (classify (-5))
    putStrLn (classify 0)
    putStrLn (classify 7)

    section "guards"
    putStrLn $ "grade 95 = " ++ [grade 95]
    putStrLn $ "grade 82 = " ++ [grade 82]
    putStrLn $ "grade 45 = " ++ [grade 45]

    section "case expression"
    putStrLn $ describe (Left 42)
    putStrLn $ describe (Right "hello")

    section "recursion (countdown)"
    countdown 5

    section "tail recursion"
    putStrLn $ "sum 1..10 = " ++ show (sumUpTo 10)

    section "higher-order (map)"
    let squares = map (^2) [1..5]
    putStrLn $ "squares [1..5] = " ++ show squares

    section "filter"
    putStrLn $ "evens [1..10] = " ++ show (filter even [1..10])

    section "list comprehensions"
    putStrLn $ "evens 1..20 = " ++ show evens
    putStrLn $ "pythagorean triples = " ++ show pythag

    section "fold"
    putStrLn $ "sum [1..10] = " ++ show (foldl (+) 0 [1..10])
    putStrLn $ "product [1..5] = " ++ show (foldl (*) 1 [1..5])

    section "takeWhile / dropWhile"
    putStrLn $ "takeWhile (< 5) [1..] = " ++ show (takeWhile (< 5) [1..])
    putStrLn $ "dropWhile (< 5) [1..10] = " ++ show (dropWhile (< 5) [1..10])

    section "Summary"
    putStrLn "Haskell control flow:"
    putStrLn "  - if/then/else (expression, not statement)"
    putStrLn "  - Guards for multi-way branching"
    putStrLn "  - case for pattern matching"
    putStrLn "  - Recursion for looping"
    putStrLn "  - Tail recursion for efficiency"
    putStrLn "  - map, filter, foldl, foldr for iteration"
    putStrLn "  - List comprehensions for generation"
